pkg/batchfetcher/batchfetcher.go

```diff
@@ -0,0 +1,131 @@
+package batchfetcher
+
+import (
+	"fmt"
+
+	availabilitydsl "github.com/filecoin-project/mir/pkg/availability/dsl"
+	availabilityevents "github.com/filecoin-project/mir/pkg/availability/events"
+	"github.com/filecoin-project/mir/pkg/dsl"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/availabilitypb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+// NewModule returns a new batch fetcher module.
+// The batch receives events output by the ordering protocol (e.g. ISS)
+// and relays them to the application in the same order.
+// It replaces the DeliverCert events from the input stream by the corresponding ProvideTransactions
+// that it obtains from the availability layer.
+// It keeps track of the current epoch (by observing the relayed NewEpoch events)
+// and automatically requests the transactions from the correct instance of the availability module.
+func NewModule(mc *ModuleConfig) modules.Module {
+	m := dsl.NewModule(mc.Self)
+
+	// The current epoch number, as announced by the ordering protocol.
+	epochNr := t.EpochNr(0)
+
+	// Map of delivered requests that is used to filter duplicates.
+	// TODO: Implement compaction (client watermarks) so that this map does not grow indefinitely.
+	delivered := make(map[t.ClientID]map[t.ReqNo]struct{})
+
+	// Queue of output events. It is required for buffering events being relayed
+	// in case a DeliverCert event that has arrived before has not yet been transformed to a ProvideTransactions event.
+	var output outputQueue
+
+	// The NewEpoch handler updates the current epoch number and forwards the event to the output.
+	dsl.UponEvent[*eventpb.Event_NewEpoch](m, func(newEpoch *eventpb.NewEpoch) error {
```

Comment:
I see. I think it might be a matter of personal preference. Using a wrapper definitely makes the final code look nicer, but I think that even using `dsl.UponEvent` is simple and readable enough, especially in such a simple case. Since the "idiomatic" way involves creating an extra file (or even a package?), personally I think that `dsl.UponEvent` is easier. Definitely for someone who is just starting to use the DSL modules.

If those wrappers were generated (I assume they will be after you merge your code generation update), that would be ideal. In a way, since the `NewEpoch` event is a "legacy" one, i.e., is not associated to a module, there would also be a question of where to put the wrapper...

Long story short, I'd prefer using `dsl.UponEvent` here, and only use wrappers when they are generated and only for events that have a clear association with a module.

Reply:
> I think it might be a matter of personal preference

The issue here is that sometimes we need to do some pre-processing of events (e.g., replacing the origin with the context). Hence, some events do require wrappers. Then I decided to add wrappers for all events for consistency. The other reason was to convert from protobuf types to proper go types.

> there would also be a question of where to put the wrapper

Such wrappers currently live in `pkg/dsl/events`.

In any case, this discussion is about to become irrelevant after I merge the code generation, so let's just go ahead without the wrappers in this PR. I'll mark the conversation as resolved.